### PR TITLE
Resolve for #26

### LINKED ISSUE #26: Remote Connection Log-In
### Overview

An extension to the completed #16 which allows remote connections.
When use is not able to access the remote machine, prompt them for user credentials to use.

https://stackoverflow.com/a/212673 gives examples on how you can perform user impersonation with a service controller.

### Expected Behavior

When unable to get services due to insufficient permission, prompt user for login details.
Or, when connecting to a machine which is not the local machine, ask user if they need to log in.

### Actual Behavior

If you are unable to view services on remote machine, an error occurs

### Steps To Reproduce Behavior

1. Enter the machine name of a remote machine you do not have permissions to.
2. Error occurs

### PR DESCRIPTION
**Pull request by object71**
### Packages
Added two nuget packages trough paket: CredentialManagement which provides the default credentials windows to ask the user. The second one is CommandLineParser that parses comman line arguments in a good manner and standard.

### Approach
The changes include catching the access denied error when connecting to a machine and asking the user to provide credentials - then a new process will start that impersonates (netonly) the user that has been provided. The netonly option is hard to do normally from code so I snatched one class that does that trough the Win32 library.

### Why this approach
The standard impersonate path was impossible as there are too many threads (Tasks) in the main window and impersonation is working on thread level. Except that I have seen anomalies with the tasks (not threads) as they would act different when impersonating and that means it is not reliable.

This pull is open to any modifications including simple refactorings to whole rework if there could be another more easier way.

## Issues Closed
Resolve #26

### CODE REVIEW

File: src/ServiceBouncer/MainForm.cs [hunk 1]
```diff
@@ -15,22 +18,45 @@ public partial class MainForm : Form
...
             Microsoft.Win32.SystemEvents.SessionSwitch += SessionSwitch;
+
 #if NET45
             //In NET45 startup type requires WMI, so it doesn't auto refresh
             dataGridStatupType.HeaderText = $"{dataGridStatupType.HeaderText} (No Auto Refresh)";
 #endif
         }
 
+        public MainForm(Options options)
+        {
```
Comment (BlythMeister) [inline src/ServiceBouncer/MainForm.cs]: Can you make this a single constructor?
I think some of the logic is missing from original constructor.

File: src/ServiceBouncer/MainForm.cs [hunk 2]
```diff
@@ -6,7 +6,10 @@
 using System.Threading;
 using System.Threading.Tasks;
 using System.Windows.Forms;
+using CredMan = CredentialManagement;
```
Comment (BlythMeister) [inline src/ServiceBouncer/MainForm.cs]: Can we not alias the types like this please.

File: paket.dependencies [hunk 3]
```diff
@@ -1,5 +1,7 @@
 framework: auto-detect
```
Comment (BlythMeister) [inline paket.dependencies]: With framework on auto it's adding a lot to the project file, can we specify the correct frameworks (net 4.5 and 4.6.1)

**Review by BlythMeister — CHANGES_REQUESTED** [review]
Couple of small changes please.

File: src/ServiceBouncer/MainForm.cs (hunk 1 above)
Comment (object71) [reply src/ServiceBouncer/MainForm.cs]: I can that was the initial way, but when refactoring decided to split. Will do

Comment (BlythMeister) [reply src/ServiceBouncer/MainForm.cs]: Yeah sometimes it's good to split, but I think in this case, make the main method pass the machine name into ctor. And put the logic into program.cs

File: src/ServiceBouncer/MainForm.cs (hunk 2 above)
Comment (object71) [reply src/ServiceBouncer/MainForm.cs]: Okay. Then I will remove this using and call the types by the full namespace path. (There wer conflicts)

File: paket.dependencies (hunk 3 above)
Comment (object71) [reply paket.dependencies]: Okay. Will check that out.

**Comment by object71** [comment]
Made changes - the second constructor was converted as a private method that would set options from the command line if they are available. This method is now called form the main constructor. Paket is now configured to use net45 and net461. Also removed the unneeded shorthand for the CredentialManager.

File: src/ServiceBouncer/MainForm.cs [hunk 4]
```diff
@@ -15,22 +17,37 @@ public partial class MainForm : Form
         private readonly List<ServiceViewModel> services;
         private bool isActive;
         private string machineHostname;
-        private int backgroundRefreshSeconds = 1;
+        private int backgroundRefreshSeconds;
 
-        public MainForm()
+        public MainForm(Options options = null)
```
Comment (BlythMeister) [inline src/ServiceBouncer/MainForm.cs]: Could you pass machine name in here.
If there is no options, then use the default of current machine.

**Review by BlythMeister — CHANGES_REQUESTED** [review]
Sorry didn't make myself clear on previous comments.

**Comment by object71** [comment]
Should be better now.

File: src/ServiceBouncer/MainForm.cs [hunk 5]
```diff
@@ -15,22 +17,33 @@ public partial class MainForm : Form
...
             services = new List<ServiceViewModel>();
             Microsoft.Win32.SystemEvents.SessionSwitch += SessionSwitch;
+            
 #if NET45
             //In NET45 startup type requires WMI, so it doesn't auto refresh
             dataGridStatupType.HeaderText = $"{dataGridStatupType.HeaderText} (No Auto Refresh)";
 #endif
         }
 
+        private void SetOptions(Options options)
```
Comment (BlythMeister) [inline src/ServiceBouncer/MainForm.cs]: This function is now redundant 🙂

File: src/ServiceBouncer/MainForm.cs [hunk 6]
```diff
@@ -15,22 +17,33 @@ public partial class MainForm : Form
...
+        private int backgroundRefreshSeconds;
 
-        public MainForm()
+        public MainForm(string machine = null)
         {
             InitializeComponent();
             isActive = true;
-            machineHostname = Environment.MachineName;
+            backgroundRefreshSeconds = 1;
+            machineHostname = machine ?? Environment.MachineName;
```
Comment (BlythMeister) [inline src/ServiceBouncer/MainForm.cs]: Pass the environment.machine into the constructor.
And then the ctor doesn't need an optional value

**Review by BlythMeister — CHANGES_REQUESTED** [review]
Sorry 😞
Thanks for sticking with this!

**Comment by object71** [comment]
Had to change the options model a bit as it cannot accept a default that isn't a constant. machine is now being passed as soon as it is parsed from the command line.

I will stick with this as much as I have to, no need to be sorry. It is me who needs to learn to type better code.

**Comment by BlythMeister** [comment]
This looks good, I want to test before merge though 👍

**Comment by object71** [comment]
No problem. Always better safe than sorry.

**Review by BlythMeister — APPROVED** [review]

**Pull request merged**